homeassistant/components/august/__init__.py

```diff
@@ -226,6 +232,29 @@ def locks(self):
         """Return a list of locks."""
         return self._locks
 
+    def api(self):
+        """Call the august api after making sure the token is up to date."""
+
+        # Check to see if we need to refresh the access token
+        # before making an api request
+        self.refresh_access_token_if_needed()
+        return self._api
+
+    def refresh_access_token_if_needed(self):
+        """Refresh the august access token if needed."""
+
+        if self._authenticator.should_refresh():
```

Comment:
I don't see a lock here, which means that you can fetch multiple new tokens at the same time.

Reply:
@balloob This probably needs a look.  Which locking mechanism is recommended?

Reply:
We only use async locks because normal locks can starve our thread pool. 

In async land we cannot do I/O as that halts the system. You can run functions that do I/O in async world with `await hass.async_add_executor_job(function, arg1, arg2)`. Functions `lock`, `unlock` and `update` need to be updated to be async `async def async_X`

Reply:
Since this can come in from doorbell, camera, or lock it looks like I'd have to convert all of these to async.   I'm still learning the codebase so its a bit outside of my comfort zone.  Since we know we have to call the activity api, I'm going to move the token refresh there since it will already be blocked from happening at the same time by the Throttle